src/OpenTelemetry/Internal/WildcardHelper.cs

```diff
@@ -14,13 +14,19 @@
 // limitations under the License.
 // </copyright>
 
+#nullable enable
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
 using System.Text.RegularExpressions;
 
 namespace OpenTelemetry;
 
 internal static class WildcardHelper
 {
-    public static bool ContainsWildcard(string value)
+    public static bool ContainsWildcard(
+        [NotNullWhen(true)]
```

Comment:
This seems a bit odd. Isn't this attribute mainly meant for `out` parameters?

Reply:
It is a bit odd, ya. Definitely more common to see it on "Try" methods with `out`s. But it is actually providing a _little_ value here.

Consider these helpers...

```csharp
public static class Helpers
{
    public static void EnsureValid(string? value)
    {
        if (string.IsNullOrEmpty(value))
            throw new ArgumentNullException(nameof(value));
    }

    public static void EnsureValidWithAnnotations([NotNull] string? value)
    {
        if (string.IsNullOrEmpty(value))
            throw new ArgumentNullException(nameof(value));
    }

    public static int DoWork(string value)
    {
        return value.Length * 18;
    }
}
```

Called like this:

```csharp
public static class Logic
{
    public static void MethodA(string? userInput)
    {
        Helpers.EnsureValid(userInput);

        // Generates CS8604: Possible null reference argument for parameter 'value' in 'int Helpers.DoWork(string value)'
        Helpers.DoWork(userInput);
    }

    public static void MethodB(string? userInput)
    {
        Helpers.EnsureValidWithAnnotations(userInput);

        // No warning
        Helpers.DoWork(userInput);
    }
}
```

Most (all?) of these nullable annotations are about hinting to the analysis engine about what was validated and known to be `null` or a valid ref so that downstream code can be flagged when it tries to de-reference something.

Here is a similar example using the wildcard annotation here:

```csharp
public static class Helpers
{
    public static bool ContainsWildcard(string? value)
    {
        if (value == null)
        {
            return false;
        }

        return value.Contains('*') || value.Contains('?');
    }

    public static bool ContainsWildcardWithAnnotations(
        [NotNullWhen(true)]
        string? value)
    {
        if (value == null)
        {
            return false;
        }

        return value.Contains('*') || value.Contains('?');
    }
}

public static class Logic
{
    public static void MethodA(string? userInput)
    {
        if (Helpers.ContainsWildcard(userInput))
        {
            // Generates CS8602: Dereference of a possibly null reference.
            var length = userInput.Length;
        }
    }

    public static void MethodB(string? userInput)
    {
        if (Helpers.ContainsWildcardWithAnnotations(userInput))
        {
            // No warning
            var length = userInput.Length;
        }
    }
}
```

So the `[NotNullWhen(true)]` annotation is just hinting to the engine that if `ContainsWildcardWithAnnotations` returns `true` then whatever string was passed in wasn't `null`. Kind of neat!

Reply:
Thanks for that explanation! The only issue is that we don't use it the way you explained. We are not referencing the input string passed to `ContainsWildcard` after we call the method, so there is no code analysis benefit coming out of it.

It's probably good for future changes (if at all we start referencing the input string passed to `ContainsWildcard` after calling the method)?

Reply:
> It's probably good for future changes (if at all we start referencing the input string passed to ContainsWildcard after calling the method)?

Ya who knows what will happen in the future. Doing this exercise I'm looking at the method in isolation. Reading the code, it is validating not-null so I added the attribute to capture that. If anything, it is good for self-documentation 🤣